lib/handoff/transfer_list.c

```diff
@@ -0,0 +1,334 @@
+	tl->hdr_size = sizeof(*tl);
+	tl->alignment = TRANSFER_LIST_INIT_MAX_ALIGN; // initial max align
+	tl->size = sizeof(*tl); // initial size is the size of header
+	tl->max_size = max_size;
+	transfer_list_update_checksum(tl);
+	return tl;
+}
+
+// 2.4.6 Relocating a TL
+bool transfer_list_relocate(struct transfer_list_header *tl, void **p, size_t max_size)
+{
+	if (!tl || !p || !(*p) || max_size == 0)
+		return false;
+
+	uintptr_t mask = (1 << tl->alignment) - 1;
+	uintptr_t new_tl_base = ((uintptr_t)(*p) + mask) & ~mask;
+
+	if (new_tl_base < (uintptr_t)(*p))
+		new_tl_base += mask + 1;
+
+	if (new_tl_base - (uintptr_t)(*p) + tl->size > max_size)
+		return false;
+
+	memmove((void *)new_tl_base, (void *)tl, tl->size);
+	((struct transfer_list_header *)new_tl_base)->max_size = max_size; //to-check
+	*p = (void *)new_tl_base;
+	transfer_list_update_checksum((struct transfer_list_header *)(*p));
+	return true;
+}
+
+static bool check_header(struct transfer_list_header *tl)
+{
+	if (!tl)
+		return false;
+
+	if (tl->signature != TRANSFER_LIST_SIGNATURE) {
+		ERROR("Bad transfer list signature %#"PRIx32"\n",
+		      tl->signature);
+		return false;
+	}
+	if (tl->version != TRANSFER_LIST_VERSION) {
+		ERROR("Unsupported transfer list version %#"PRIx32"\n",
+		      tl->version);
+		return false;
+	}
+	if (!tl->max_size) {
+		ERROR("Bad transfer list max size %#"PRIx32"\n",
+		      tl->max_size);
+		return false;;
+	}
+	if (tl->size > tl->max_size) {
+		ERROR("Bad transfer list size %#"PRIx32"\n", tl->size);
+		return false;;
+	}
+	return true;
+}
+
+// 2.4.1 Validating a TL header
+struct transfer_list_header *transfer_list_check_header(struct transfer_list_header *tl)
+{
+	if (!tl || !check_header(tl))
+		return NULL;
+	if (!transfer_list_verify_checksum(tl)) {
+		ERROR("Bad transfer list checksum %#x\n", tl->checksum);
+		return NULL;
+	}
+	return tl;
+}
+
+static struct transfer_list_entry *transfer_list_next(struct transfer_list_header *tl,
+					  struct transfer_list_entry *last)
+{
+	if (!tl)
+		return NULL;
+
+	struct transfer_list_entry *te = NULL;
+	uintptr_t tl_ev = (uintptr_t)tl + tl->size;
+	uintptr_t va = 0;
+	uintptr_t ev = 0;
+	size_t sz = 0;
+
+	if (last) {
+		va = (uintptr_t)last;
+		if (ADD_OVERFLOW(last->hdr_size, last->data_size, &sz) ||
+		    ADD_OVERFLOW(va, sz, &va) ||
+		    ROUNDUP_OVERFLOW(va, (1 << tl->alignment), &va))
+			return NULL;
+	} else {
+		va = (uintptr_t)(tl + 1);
+	}
+	te = (struct transfer_list_entry *)va;
+
+	if (va + sizeof(*te) > tl_ev || te->hdr_size < sizeof(*te) ||
+	    ADD_OVERFLOW(va, te->hdr_size, &ev) ||
+	    ADD_OVERFLOW(ev, te->data_size, &ev) ||
+	    ev > tl_ev)
+		return NULL;
+
+	return te;
+}
+
+static uint8_t calc_checksum(struct transfer_list_header *tl)
+{
+	if (!tl)
+		return 0;
+
+	uint8_t *b = (uint8_t *)tl;
+	uint8_t cs = 0;
+	size_t n = 0;
+
+	for (n = 0; n < tl->size; n++)
+		cs += b[n];
+
+	return cs;
+}
+
+void transfer_list_update_checksum(struct transfer_list_header *tl)
```

Comment:
Dpesn't TF-A have checksum calculating functions?

Reply:
Did not find any same checksum implementation in lib or plat common code.